File: node/mods/rustscript/lib/opcodes/setarray.js
/**
 * Purpose: SETARRAY opcode — replace a context location with a deep-cloned array.
 *
 * Language shape:
 *   script.destination  — context.* path to overwrite
 *   script.source       — array (resolve_ref or tx collection ref)
 *
 * Special unresolved sources: tx.from / tx.to / tx.path / tx.from.p2sh / tx.to.p2sh
 * resolve to the corresponding collections (not their lengths).
 *
 * Read-only destinations (script.*, witness.*, tx.*, blk.*) are rejected.
 */

module.exports = {
  name: 'SETARRAY',
  description: `
Replaces a writable context location with a deep clone of a source array.

destination must be a context.* path. source is resolved via normal RustScript
references, or as a transaction collection (tx.from, tx.to, tx.path,
tx.from.p2sh, tx.to.p2sh) when unresolved.

The source is not modified.
`,
  exampleScript: {
    op: 'SETARRAY',
    destination: 'context.successors',
    source: '__opcodes.importarray.successors'
  },
  schema: {
    destination: 'string',
    source: 'reference'
  },

  execute(node, context) {
    if (!node || typeof node !== 'object' || !context || typeof context !== 'object') {
      return false;
    }

    const destination = node.destination;
    if (typeof destination !== 'string' || destination.length === 0) {
      return false;
    }

    if (!Object.prototype.hasOwnProperty.call(node, 'source')) {
      return false;
    }

    if (!destination.startsWith('context.')) {
      return false;
    }

    const path = destination.slice('context.'.length);
    if (!path || isForbiddenWritePath(path)) {
      return false;
    }

    const sourceArray = resolveSourceArray(node.source, context);
    if (!Array.isArray(sourceArray)) {
      return false;
    }

    return setContextPath(context, path, deepClone(sourceArray));
  }
};

function isForbiddenWritePath(path) {
  return (
    path === 'script' ||
    path.startsWith('script.') ||
    path === 'witness' ||
    path.startsWith('witness.') ||
    path === 'tx' ||
    path.startsWith('tx.') ||
    path === 'blk' ||
    path.startsWith('blk.')
  );
}

function pathSegments(path) {
  return String(path)
    .replace(/\[/g, '.')
    .replace(/\]/g, '')
    .split('.')
    .filter((s) => s.length > 0);
}

function setContextPath(root, path, value) {
  const segments = pathSegments(path);
  if (!segments.length) {
    return false;
  }

  let current = root;
  for (let i = 0; i < segments.length; i += 1) {
    const segment = segments[i];
    const isLast = i === segments.length - 1;
    const asIndex = /^[0-9]+$/.test(segment) ? Number(segment) : null;

    if (isLast) {
      if (asIndex !== null) {
        if (!Array.isArray(current) || asIndex >= current.length) {
          return false;
        }
        current[asIndex] = value;
        return true;
      }
      if (!current || typeof current !== 'object' || Array.isArray(current)) {
        return false;
      }
      current[segment] = value;
      return true;
    }

    if (asIndex !== null) {
      if (!Array.isArray(current) || asIndex >= current.length) {
        return false;
      }
      current = current[asIndex];
      continue;
    }

    if (!current || typeof current !== 'object' || Array.isArray(current)) {
      return false;
    }
    if (!Object.prototype.hasOwnProperty.call(current, segment)) {
      current[segment] = {};
    }
    current = current[segment];
  }

  return false;
}

function isP2shSlip(slip) {
  if (!slip || typeof slip !== 'object') {
    return false;
  }
  if (slip.slip_type === 'Bound' || slip.slip_type === 9) {
    return false;
  }
  const pk = slip.public_key || slip.publicKey;
  if (Array.isArray(pk) && pk[0] === 0) {
    return true;
  }
  if (pk && typeof pk === 'object' && pk[0] === 0) {
    return true;
  }
  return false;
}

function resolveSourceArray(source, context) {
  const resolved = resolveRef(context, source);
  if (Array.isArray(resolved)) {
    return resolved;
  }

  if (typeof source !== 'string') {
    return null;
  }

  if (!context.tx || typeof context.tx !== 'object') {
    if (source === 'tx.from' || source === 'tx.to' || source === 'tx.path'
      || source === 'tx.from.p2sh' || source === 'tx.to.p2sh') {
      return null;
    }
  }

  if (source === 'tx.from') {
    return Array.isArray(context.tx.from) ? context.tx.from : null;
  }
  if (source === 'tx.to') {
    return Array.isArray(context.tx.to) ? context.tx.to : null;
  }
  if (source === 'tx.path') {
    return Array.isArray(context.tx.path) ? context.tx.path : null;
  }
  if (source === 'tx.from.p2sh') {
    return Array.isArray(context.tx.from) ? context.tx.from.filter(isP2shSlip) : null;
  }
  if (source === 'tx.to.p2sh') {
    return Array.isArray(context.tx.to) ? context.tx.to.filter(isP2shSlip) : null;
  }

  return null;
}

function deepClone(value) {
  return JSON.parse(JSON.stringify(value));
}

function resolveRef(root, ref) {
  if (typeof ref !== 'string') {
    return ref;
  }
  const parts = ref.split('.');
  let cursor = root;
  for (let i = 0; i < parts.length; i += 1) {
    const key = parts[i];
    if (
      !cursor ||
      typeof cursor !== 'object' ||
      !Object.prototype.hasOwnProperty.call(cursor, key)
    ) {
      return ref;
    }
    cursor = cursor[key];
  }
  return cursor;
}

File: rust/saito-core/src/core/consensus/scripting/opcodes/setarray.rs
use crate::core::consensus::block::Block;
use crate::core::consensus::slip::{Slip, SlipType};
use crate::core::consensus::transaction::Transaction;
use serde_json::Value;

use super::super::script::resolve_ref;

/// SETARRAY — replace a context location with a deep clone of a source array.
///
/// Script shape:
/// ```json
/// {
///   "op": "SETARRAY",
///   "destination": "context.successors",
///   "source": "__opcodes.importarray.successors"
/// }
/// ```
///
/// `source` is resolved via resolve_ref. If unresolved, the special collection
/// refs `tx.from` / `tx.to` / `tx.path` / `tx.from.p2sh` / `tx.to.p2sh` resolve
/// to the corresponding arrays.
///
/// Destinations must be `context.*` paths. Writes to `script.*`, `witness.*`,
/// `tx.*`, and `blk.*` are rejected.
pub struct SetArray {
    pub name: String,
    pub description: String,
    pub script: String,
    pub schema: Value,
}

fn is_forbidden_write_path(path: &str) -> bool {
    path == "script"
        || path.starts_with("script.")
        || path == "witness"
        || path.starts_with("witness.")
        || path == "tx"
        || path.starts_with("tx.")
        || path == "blk"
        || path.starts_with("blk.")
}

/// Normalize `foo[0].bar` into dotted segments compatible with resolve_ref lookup.
fn path_segments(path: &str) -> Vec<String> {
    let normalized = path.replace('[', ".").replace(']', "");
    normalized
        .split('.')
        .filter(|s| !s.is_empty())
        .map(|s| s.to_string())
        .collect()
}

fn set_context_path(root: &mut Value, path: &str, value: Value) -> bool {
    let segments = path_segments(path);
    if segments.is_empty() {
        return false;
    }

    let last = segments.len() - 1;
    let mut current = root;

    for (i, segment) in segments.iter().enumerate() {
        if i == last {
            if let Ok(index) = segment.parse::<usize>() {
                let Some(arr) = current.as_array_mut() else {
                    return false;
                };
                if index >= arr.len() {
                    return false;
                }
                arr[index] = value;
                return true;
            }

            if !current.is_object() {
                *current = Value::Object(serde_json::Map::new());
            }
            if let Value::Object(map) = current {
                map.insert(segment.clone(), value);
                return true;
            }
            return false;
        }

        if let Ok(index) = segment.parse::<usize>() {
            let Some(arr) = current.as_array_mut() else {
                return false;
            };
            if index >= arr.len() {
                return false;
            }
            current = &mut arr[index];
            continue;
        }

        if !current.is_object() {
            *current = Value::Object(serde_json::Map::new());
        }

        let Value::Object(map) = current else {
            return false;
        };

        if !map.contains_key(segment) {
            map.insert(segment.clone(), Value::Object(serde_json::Map::new()));
        }

        current = map.get_mut(segment).unwrap();
    }

    false
}

/// P2SH slip filter matching resolve_p2sh_slip_field / ARRAYIFY.
fn is_p2sh_slip(slip: &Slip) -> bool {
    if slip.slip_type == SlipType::Bound {
        return false;
    }
    slip.public_key[0] == 0x00
}

fn slips_to_value(slips: &[Slip]) -> Option<Value> {
    serde_json::to_value(slips).ok()
}

fn p2sh_slips_to_value(slips: &[Slip]) -> Option<Value> {
    let filtered: Vec<&Slip> = slips.iter().filter(|s| is_p2sh_slip(s)).collect();
    serde_json::to_value(filtered).ok()
}

fn resolve_source_array(
    source: &Value,
    context: &Value,
    tx: Option<&Transaction>,
    blk: Option<&Block>,
) -> Option<Value> {
    let resolved = resolve_ref(source, context, tx, blk);
    if resolved.is_array() {
        return Some(resolved);
    }

    //
    // special collection refs — only after normal resolve_ref fails to yield
    // an array
    //
    let Some(s) = source.as_str() else {
        return None;
    };

    match s {
        "tx.from" => tx.and_then(|t| slips_to_value(&t.from)),
        "tx.to" => tx.and_then(|t| slips_to_value(&t.to)),
        "tx.path" => tx.and_then(|t| serde_json::to_value(&t.path).ok()),
        "tx.from.p2sh" => tx.and_then(|t| p2sh_slips_to_value(&t.from)),
        "tx.to.p2sh" => tx.and_then(|t| p2sh_slips_to_value(&t.to)),
        _ => None,
    }
}

impl SetArray {
    pub fn validate(context: &mut Value, tx: Option<&Transaction>, blk: Option<&Block>) -> u8 {
        let destination = context["script"]["destination"]
            .as_str()
            .unwrap_or("")
            .to_string();
        if destination.is_empty() {
            return 0;
        }

        let Some(path) = destination.strip_prefix("context.") else {
            return 0;
        };
        let path = path.to_string();
        if path.is_empty() || is_forbidden_write_path(&path) {
            return 0;
        }

        if context["script"].get("source").is_none() {
            return 0;
        }

        let Some(source_array) =
            resolve_source_array(&context["script"]["source"].clone(), context, tx, blk)
        else {
            return 0;
        };

        if !source_array.is_array() {
            return 0;
        }

        // Value::clone is a deep copy of the JSON tree
        let cloned = source_array.clone();

        if set_context_path(context, &path, cloned) {
            1
        } else {
            0
        }
    }
}
